src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -17,64 +18,108 @@ internal CoseSign1Message(CoseHeaderMap protectedHeader, CoseHeaderMap unprotect
-            byte[] toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtectedHeader, content);
-            byte[] signature = SignWithRSA(key, toBeSigned, hashAlgorithm);
-            return SignCore(encodedProtectedHeader, GetEmptyCborMap(), signature, content, isDetached);
-        }
+            => SignCore(content.AsSpan(), key, hashAlgorithm, KeyType.RSA, null, null, isDetached);
 
         [UnsupportedOSPlatform("browser")]
         public static byte[] Sign(byte[] content!!, AsymmetricAlgorithm key!!, HashAlgorithmName hashAlgorithm, CoseHeaderMap? protectedHeaders = null, CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false)
+            => SignCore(content.AsSpan(), key, hashAlgorithm, GetKeyType(key), protectedHeaders, unprotectedHeaders, isDetached);
+
+        [UnsupportedOSPlatform("browser")]
+        public static byte[] Sign(ReadOnlySpan<byte> content, AsymmetricAlgorithm key!!, HashAlgorithmName hashAlgorithm, CoseHeaderMap? protectedHeaders = null, CoseHeaderMap? unprotectedHeaders = null, bool isDetached = false)
+            => SignCore(content, key, hashAlgorithm, GetKeyType(key), protectedHeaders, unprotectedHeaders, isDetached);
+
+        [UnsupportedOSPlatform("browser")]
+        internal static byte[] SignCore(ReadOnlySpan<byte> content, AsymmetricAlgorithm key, HashAlgorithmName hashAlgorithm, KeyType keyType, CoseHeaderMap? protectedHeaders, CoseHeaderMap? unprotectedHeaders, bool isDetached)
         {
-            KeyType keyType = key switch
+            ValidateBeforeSign(protectedHeaders, unprotectedHeaders, keyType, hashAlgorithm, out int? algHeaderValueToSlip);
+
+            int expectedSize = ComputeEncodedSize(protectedHeaders, unprotectedHeaders, algHeaderValueToSlip, content.Length, isDetached, key.KeySize, keyType);
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(expectedSize);
+
+            try
             {
-                ECDsa => KeyType.ECDsa,
-                RSA => KeyType.RSA,
-                _ => throw new CryptographicException(SR.Format(SR.Sign1SignUnsupportedKey, key.GetType()))
-            };
+                int bytesWritten = CreateCoseSign1Message(content, rentedBuffer, key, hashAlgorithm, protectedHeaders, unprotectedHeaders, isDetached, algHeaderValueToSlip, keyType);
+                Debug.Assert(expectedSize == bytesWritten);
 
-            ThrowIfDuplicateLabels(protectedHeaders, unprotectedHeaders);
+                return rentedBuffer.AsSpan(0, bytesWritten).ToArray();
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer, clearArray: true);
+            }
+        }
 
-            int? algHeaderValueToSlip = ValidateOrSlipAlgorithmHeader(protectedHeaders, unprotectedHeaders, keyType, hashAlgorithm);
+        [UnsupportedOSPlatform("browser")]
+        public static bool TrySign(
+            ReadOnlySpan<byte> content, Span<byte> destination,
```

Comment:
We don't have a written repository rule here, but the general rule seems to be one of "all parameters on the same line as the method name" or "one parameter per line".  I don't think we have much in the way of hodgepodging.

Reply:
I prefer "all parameters on the same line" but last time I got feedback about signatures being too large for GitHub file viewer.

Reply:
Sure, and since the signature here is long it should be chopped to one parameter per line :smile: